### PR TITLE
[server] disallow deauthorization if account is blocked

### PR DESCRIPTION
**Pull request by geropl**
## Description
Note: This is already deployed in prod.

## Related Issue(s)

Fixes #

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by jankeromnes** [comment]
Will take a look 👀

File: components/server/src/auth/authenticator.ts [hunk 1]
```diff
@@ -151,7 +151,7 @@ export class Authenticator {
 
     async deauthorize(req: express.Request, res: express.Response, next: express.NextFunction) {
         const user = req.user;
-        if (!req.isAuthenticated() || !User.is(user)) {
+        if (!req.isAuthenticated() || !User.is(user) || user.blocked) {
             log.info({ sessionId: req.sessionID }, `User is not authenticated.`);
             res.redirect(this.getSorryUrl(`Not authenticated. Please login.`));
```
Comment (jankeromnes) [inline components/server/src/auth/authenticator.ts:156]: This is not the correct failure mode for blocked users, right?

**Review by jankeromnes — COMMENTED** [review]
The `checkAndBlockUser` changes mostly look fine to me.

However, I'm unsure about the `deauthorize` changes. Seems a bit strange to tell logged in but blocked users to please log in. 😬 Also -- what is the benefit of preventing blocked users from disconnecting their Integrations?

**Comment by geropl** [comment]
> what is the benefit of preventing blocked users from disconnecting their Integrations?

@jankeromnes That's the point of this PR - blocking blocked users from removeing their identities. :upside_down_face:

File: components/server/src/auth/authenticator.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/auth/authenticator.ts:156]: > This is not the correct failure mode for blocked users, right?

Nope, but I think we have bigger problems atm then caring about UX for blocked users.

Let's take and note/follow up PR and merge this one asap. :pray:

Comment (jankeromnes) [reply components/server/src/auth/authenticator.ts:156]: I wonder if a "cleaner" location for this fix might be:

https://github.com/gitpod-io/gitpod/blob/ed30d9696fd1c0ca5775a2d44d6b4a9e5709ba63/components/server/src/user/user-controller.ts#L147-L150

**Comment by jankeromnes** [comment]
@geropl Do you think [this commit](https://github.com/gitpod-io/gitpod/commit/3b806e8edb61d57cbd51327e5cf6f8112afc49ec) could work? Should I push it to this PR?

I think this would be a bit cleaner. (It's not about blocked user UX, but keeping consistent failure modes.)

**Comment by jankeromnes** [comment]
Works "as intended", but the failure mode looks a bit surprising:

| De-authorizing shows inaccurate error | Deleting shows nothing | Authorizing opens a silly pop-up |
| --- | --- | --- |
| Screenshot 2022-04-19 at 14 47 24 | Screenshot 2022-04-19 at 14 47 51 | Screenshot 2022-04-19 at 14 49 34 |

**Review by jankeromnes — APPROVED** [review]
PR looks good as is. 👍 Leaving the aforementioned [UX problems](https://github.com/gitpod-io/gitpod/pull/9372#issuecomment-1102610850) as follow-ups.

**Pull request merged**